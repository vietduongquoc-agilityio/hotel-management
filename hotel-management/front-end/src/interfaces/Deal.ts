export interface DealData {
  dealName: string;
  referenceNumber: string;
  startDate: Date;
  endDate: Date;
  roomType: string;
  statusDeal: string;
  reservationsLeft: number;
  documentId: string;
}

export type NewDealData = Omit<DealData, "documentId">;
