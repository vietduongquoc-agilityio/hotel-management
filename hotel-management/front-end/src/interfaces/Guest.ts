export interface GuestData {
    guestName: string;
    roomType: string;
    stay: number;
    price: number;
    registrationNumber: string;
    totalAmount: number;
    checkInDate: Date;
    documentId: string;
}

export type NewGuestData = Omit<GuestData, "documentId">;
