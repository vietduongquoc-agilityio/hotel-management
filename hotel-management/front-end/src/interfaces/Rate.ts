export interface RateData {
  roomType: string;
  cancellationPolicy: string;
  deals: string;
  dealPrice: string;
  availability: string;
  documentId: string;
  rate: string;
  totalOfBooked?: number;
}

export type NewRateData = Omit<RateData, "documentId">;
