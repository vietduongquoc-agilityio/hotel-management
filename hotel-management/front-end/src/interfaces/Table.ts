export interface EditFunctionType<T> {
  initialData: T;
  onEditFunction: (updatedData: T) => void;
}

export interface DeleteFunctionType {
  documentId: string;
  onDeleteFunction: (id: string) => void;
}
