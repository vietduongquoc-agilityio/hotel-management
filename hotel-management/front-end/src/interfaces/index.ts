export * from "./Rate";
export * from "./Room";
export * from "./Table";
export * from "./Guest"
