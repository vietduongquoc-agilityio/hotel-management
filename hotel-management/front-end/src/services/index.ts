export * from "./Rate/rateServices";
export * from "./Room/roomService";
