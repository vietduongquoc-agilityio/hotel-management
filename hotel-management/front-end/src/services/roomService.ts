import axios from "axios";

//InterFace
import { NewRoomData } from "@/Interface/Room";

const BASE_URL = process.env.VITE_BASE_URL;

export const getRooms = async (
  page: number,
  pageSize: number,
  field?: string,
  value?: string
) => {
  try {
    const filters = field && value ? { [`filters[${field}]`]: value } : {};
    const response = await axios.get(`${BASE_URL}/rooms`, {
      params: {
        ...filters,
        "sort[0]": "createdAt:desc",
        "pagination[page]": page,
        "pagination[pageSize]": pageSize,
      },
    });

    if (response.data && response.data.data && response.data.meta.pagination) {
      return {
        rooms: response.data.data,
        pagination: response.data.meta.pagination,
      };
    }
    return response.data;
  } catch (error) {
    console.error("Error fetching room data", error);
    throw error;
  }
};

export const createRoomApi = async (roomData: NewRoomData) => {
  try {
    const response = await axios.post(`${BASE_URL}/rooms`, {
      data: roomData,
    });
    return response.data;
  } catch (error) {
    console.error("Error in createRoom:", error);
    throw error;
  }
};

export const getRoomById = (roomId: string) =>
  axios.get(`${BASE_URL}/rooms/${roomId}`).then((response) => response.data);

export const updateRoom = (roomId: string, roomData: NewRoomData) => {
  if (!roomId) throw new Error("Missing document ID for room update.");
  return axios
    .put(`${BASE_URL}/rooms/${roomId}`, { data: roomData })
    .then((response) => response.data);
};

export const deleteRoom = async (roomId: string) => {
  const response = await axios.delete(`${BASE_URL}/rooms/${roomId}`);
  return response.data;
};
