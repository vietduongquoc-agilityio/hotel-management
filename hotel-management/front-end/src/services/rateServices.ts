/* eslint-disable @typescript-eslint/no-unused-vars */
import axios from "axios";

// InterFace
import { NewRateData } from "@/interfaces/Rate";

const BASE_URL = process.env.VITE_BASE_URL;
// Rate Service
export const getRates = async (page: number, pageSize: number) => {
  try {
    const response = await axios.get(`${BASE_URL}/rates`, {
      params: {
        "sort[0]": "createdAt:desc",
        "pagination[page]": page,
        "pagination[pageSize]": pageSize,
      },
    });
    return response.data;
  } catch (error) {
    console.error("Error fetching rate data", error);
    throw error;
  }
};

// Check the list of rooms using a `rate` tool
export const getRoomsUsingRate = async (rateId: string) => {
  try {
    const response = await axios.get(`${BASE_URL}/rooms`, {
      params: { rateId },
    });
    return response.data;
  } catch (error) {
    console.error("Error fetching rooms using rate", error);
    throw error;
  }
};

export const createRateApi = async (rateData: NewRateData) => {
  try {
    const response = await axios.post(`${BASE_URL}/rates`, {
      data: rateData,
    });
    return response.data;
  } catch (error) {
    console.error("Error in createRate:", error);
    throw error;
  }
};

export const updateRate = (rateId: string, rateData: NewRateData) => {
  if (!rateId) throw new Error("Missing document ID for rate update.");
  return axios
    .put(`${BASE_URL}/rates/${rateId}`, { data: rateData })
    .then((response) => response.data);
};

export const deleteRate = async (rateId: string) => {
  const response = await axios.delete(`${BASE_URL}/rates/${rateId}`);
  return response.data;
};
