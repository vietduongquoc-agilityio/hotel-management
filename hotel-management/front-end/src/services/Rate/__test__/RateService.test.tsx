import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { getRates, createRateApi, updateRate, deleteRate } from "@/services";
import { NewRateData } from "@/interfaces";

jest.mock("@chakra-ui/react", () => ({
  createStandaloneToast: jest.fn(() => ({
    toast: jest.fn(),
  })),
}));

const mock = new MockAdapter(axios);
const BASE_URL = process.env.VITE_BASE_URL;

describe("rateService", () => {
  afterEach(() => {
    mock.reset();
    jest.clearAllMocks();
  });

  describe("getRates", () => {
    it("should fetch rates with pagination and match snapshot", async () => {
      const mockResponse = {
        data: [
          { id: "1", attributes: { name: "Rate 1", price: 100 } },
          { id: "2", attributes: { name: "Rate 2", price: 150 } },
        ],
        meta: { pagination: { page: 1, pageSize: 10, pageCount: 1, total: 2 } },
      };

      mock.onGet(`${BASE_URL}/rates`).reply(200, mockResponse);

      const result = await getRates(1, 10);
      expect(result).toMatchSnapshot();
    });
  });

  describe("createRateApi", () => {
    it("should create a new rate and match snapshot", async () => {
      const rateData: NewRateData = {
        roomType: "Deluxe",
        cancellationPolicy: "Flexible",
        deals: "None",
        dealPrice: "0",
        rate: "200",
        availability: "5",
        totalOfRooms: 0,
        totalOfBooked: 0,
      };
      const mockResponse = { data: { id: "1", attributes: { ...rateData } } };

      mock.onPost(`${BASE_URL}/rates`).reply(200, mockResponse);
    });

    it("should show error toast on failure", async () => {
      const rateData: NewRateData = {
        roomType: "Deluxe",
        cancellationPolicy: "Flexible",
        deals: "None",
        dealPrice: "0",
        rate: "200",
        availability: "5",
        totalOfRooms: 0,
        totalOfBooked: 0,
      };
      mock.onPost(`${BASE_URL}/rates`).reply(500);

      await expect(createRateApi(rateData)).rejects.toThrow(
        "Request failed with status code 500"
      );
    });
  });

  describe("updateRate", () => {
    it("should update a rate by ID and match snapshot", async () => {
      const rateId = "1";
      const rateData: NewRateData = {
        roomType: "Updated Deluxe",
        cancellationPolicy: "Moderate",
        deals: "Spring Discount",
        dealPrice: "20",
        rate: "180",
        availability: "3",
        totalOfRooms: 0,
        totalOfBooked: 0,
      };
      const mockResponse = {
        data: { id: rateId, attributes: { ...rateData } },
      };

      mock.onPut(`${BASE_URL}/rates/${rateId}`).reply(200, mockResponse);
    });

    it("should throw error if rateId is missing", async () => {
      const rateData: NewRateData = {
        roomType: "Updated Deluxe",
        cancellationPolicy: "Moderate",
        deals: "Spring Discount",
        dealPrice: "20",
        rate: "180",
        availability: "3",
        totalOfRooms: 0,
        totalOfBooked: 0,
      };

      await expect(updateRate("", rateData)).rejects.toThrow(
        "Missing document ID for rate update."
      );
    });

    it("should show error toast on update failure", async () => {
      const rateId = "1";
      const rateData: NewRateData = {
        roomType: "Updated Deluxe",
        cancellationPolicy: "Moderate",
        deals: "Spring Discount",
        dealPrice: "20",
        rate: "180",
        availability: "3",
        totalOfRooms: 0,
        totalOfBooked: 0,
      };

      mock.onPut(`${BASE_URL}/rates/${rateId}`).reply(500);

      await expect(updateRate(rateId, rateData)).rejects.toThrow(
        "Request failed with status code 500"
      );
    });
  });

  describe("deleteRate", () => {
    it("should delete a rate by ID and match snapshot", async () => {
      const rateId = "1";
      const mockResponse = { message: "Rate deleted successfully" };

      mock.onDelete(`${BASE_URL}/rates/${rateId}`).reply(200, mockResponse);

      const result = await deleteRate(rateId);
      expect(result).toMatchSnapshot();
    });

    it("should show error toast on delete failure", async () => {
      const rateId = "1";
      mock.onDelete(`${BASE_URL}/rates/${rateId}`).reply(404);

      await expect(deleteRate(rateId)).rejects.toThrow(
        "Request failed with status code 404"
      );
    });
  });
});
