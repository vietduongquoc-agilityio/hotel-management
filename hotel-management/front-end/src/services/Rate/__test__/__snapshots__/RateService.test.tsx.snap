// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rateService deleteRate should delete a rate by ID and match snapshot 1`] = `
{
  "message": "Rate deleted successfully",
}
`;

exports[`rateService getRates should fetch rates with pagination and match snapshot 1`] = `
{
  "data": [
    {
      "attributes": {
        "name": "Rate 1",
        "price": 100,
      },
      "id": "1",
    },
    {
      "attributes": {
        "name": "Rate 2",
        "price": 150,
      },
      "id": "2",
    },
  ],
  "meta": {
    "pagination": {
      "page": 1,
      "pageCount": 1,
      "pageSize": 10,
      "total": 2,
    },
  },
}
`;
